Add support for Sensor state class

frenck opened the pull request:
# What does this implement/fix? 

Add support for setting a sensor `state_class` to Sensor entities.

- `aioesphomeapi`: <https://github.com/esphome/aioesphomeapi/pull/33>
- Home Assistant Core: <https://github.com/home-assistant/core/pull/51090>
- ESPHome documentation: <https://github.com/esphome/esphome-docs/pull/1201>

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuration files to keep working)

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1201
  
# Test Environment

- [x] ESP32
- [ ] ESP8266
- [ ] Windows
- [x] Mac OS
- [x] Linux

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
sensor:
  - platform: template
    state_class: measurement
    ...
```

# Explain your changes

Describe your changes here to communicate to the maintainers **why we should accept this pull request**.
Very important to fill if no issue linked

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

frenck committed:
Add support for Sensor state class

balloob commented on esphome/components/api/api_pb2.cpp:
```diff
@@ -62,6 +62,16 @@ template<> const char *proto_enum_to_string<enums::FanDirection>(enums::FanDirec
+    case enums::STATE_CLASS_MEASUREMENT:
+      return "STATE_CLASS_MEASUREMENT";
+    default:
+      return "UNKNOWN";
```
Should default be none ?

frenck replied:
Generated code :)

frenck committed:
Fix state class for pulse meter

frenck committed:
Fix state class for pulse counter

frenck commented:
Converted to draft, for now, going to add defaults to everything.

frenck committed:
Set the state_class on all component sensors

frenck committed:
Set default state class to none

frenck commented:
Done

OttoWinter reviewed (commented):
Nice, the core implementation looks fine (apart from maybe allowing `none` also as an option to `state_class` to have a consistent schema in ESPHome); and that part should be ready to merge

For the adding `state_class` part of this PR: I'd really appreciate if we could have a _consistent_ definition of `measurement` state class, see comment below.

OttoWinter commented on esphome/components/gps/__init__.py:
```diff
@@ -34,22 +36,26 @@
             ),
             cv.Optional(CONF_ALTITUDE): sensor.sensor_schema(
-                UNIT_METER, ICON_EMPTY, 1, DEVICE_CLASS_EMPTY
+                UNIT_METER, ICON_EMPTY, 1, DEVICE_CLASS_EMPTY, STATE_CLASS_NONE
```
And sorry about being picky about this; There've just been too many times that ambiguous definitions/uses of a definition have caused trouble in ESPHome (meaning of white value, brightness 0 in old light model, the old climate entity model); so I'd like it to be consistent from the start

frenck replied:
Simply put: I don't care about it.
I rather approach this from a user perspective, and not from details. In this case, it is safe to assume a user isn't interested in the min/max/mean values of their long/lats in long-term stats in general.

I'm not changing this.

OttoWinter replied:
> I rather approach this from a user perspective, and not from details. In this case, it is safe to assume a user isn't interested in the min/max/mean values of their long/lats in long-term stats in general.

And that is _ok_, I completely support approaching it that way.

However, I don't want to end up with a situation where the official definition differs in a lot of places from what we're implementing here (as that is just confusing). I think the *definition* should be updated to reflect this approach (to make it match)

> To be clear, the definition can be like "use measurement to make HA aggregate statistics", but then the official definition should reflect that.

frenck replied:
> I think the definition should be updated to reflect this approach

Yeah, maybe. A large part of this all is in active development and absolutely not fully figured out. An approach is an MVP approach where we start marking measurements that are obvious and figure out a way to work with the data collected.

I'm sure that help is appreciated on the other end, so feel free to suggest or make improvements.

frenck commented on esphome/components/pulse_meter/sensor.py:
```diff
@@ -49,15 +51,15 @@ def validate_pulse_meter_pin(value):
         cv.Optional(CONF_TIMEOUT, default="5min"): validate_timeout,
         cv.Optional(CONF_TOTAL): sensor.sensor_schema(
-            UNIT_PULSES, ICON_PULSE, 0, DEVICE_CLASS_EMPTY
+            UNIT_PULSES, ICON_PULSE, 0, DEVICE_CLASS_EMPTY, STATE_CLASS_NONE
```
Incorrect, it doesn't show current measurement, but an integral. Those are supported, but require a `last_reset` attribute, which isn't in this PR thus can't be set.

frenck replied:
In the end, this still applies:

![image](https://user-images.githubusercontent.com/195327/119634427-0ad02000-be13-11eb-8a69-fbc3e1c0fda8.png)

The fact is, we don't have a `last_reset` now, which should be provided, as these sensors do accumulate.

frenck commented on esphome/components/sensor/__init__.py:
```diff
@@ -69,6 +71,14 @@
+sensor_ns = cg.esphome_ns.namespace("sensor")
+StateClasses = sensor_ns.enum("StateClass")
+STATE_CLASSES = {
+    "": StateClasses.STATE_CLASS_NONE,
```
> This is consistent with how ESPHome usually encodes "lack of" a value

While you claim it is consistent, it is actually not. The most obvious and closest example to this attribute would be `device_class`.

A lot of places use and describe an empty string. IMHO, adding 2 ways to solve the same if a bad pattern. It should either be an empty string or `none`, not both.

I suggest the ESPHome project figures out which one it should be and stick to the single one.

OttoWinter replied:
Ah yes, you're right with `device_class` also having `""`. Then it's ok.

frenck commented:
Rebased to fix merge conflicts and be up to par with the latest dev.

frenck committed:
Remove state class from energy sensors

frenck committed:
Update with latest change or misses

frenck commented:
Rebased the PR, added state classes to some new components, updated a missing one, removed some state classes from component sensors that shouldn't have a measurement qualifier yet (as last_reset is currently not supported).

Made another round through all components. I think it is good to go from my end.

jesserockz commented on esphome/components/sdm_meter/sensor.py:
```diff
@@ -74,19 +81,27 @@
             cv.Optional(CONF_PHASE_C): PHASE_SCHEMA,
             cv.Optional(CONF_FREQUENCY): sensor.sensor_schema(
-                UNIT_HERTZ, ICON_CURRENT_AC, 3
+                UNIT_HERTZ, ICON_CURRENT_AC, 3, DEVICE_CLASS_EMPTY, STATE_CLASS_NONE
```
This is a measurement in time and you might want the min/max of this
```suggestion
                UNIT_HERTZ, ICON_CURRENT_AC, 3, DEVICE_CLASS_EMPTY, STATE_CLASS_MEASUREMENT
```

jesserockz committed:
Update esphome/components/sdm_meter/sensor.py

jesserockz reviewed (approved):
Looks good from my end

jesserockz committed:
Fix formatting
